### PR TITLE
SARIF parser - add more information to findings

### PR DESCRIPTION
**Pull request by StefanFl**
see title 😃

### CODE REVIEW

**Comment by StefanFl** [comment]
@damiencarol

File: dojo/unittests/tools/test_sarif_parser.py [hunk 1]
```diff
@@ -297,25 +326,43 @@ def test_flawfinder(self):
...
+            description = '''**Result message:** buffer/sscanf:The scanf() family's %s operation, without a limit specification, permits buffer overflows (CWE-120, CWE-20).
+**Snippet:**       if (sscanf(argv[i], "%[^=]=%s", name, val) == 2) {
+**Rule name:** buffer/sscanf
+**Rule short description:** The scanf() family's %s operation, without a limit specification, permits buffer overflows (CWE-120, CWE-20).
+'''
+            self.assertEqual(description, finding.description)
             self.assertEqual("src/cli_main.cc", finding.file_path)
             self.assertEqual(482, finding.line)
             self.assertEqual("FF1021", finding.vuln_id_from_tool)
+            self.assertEqual('https://cwe.mitre.org/data/definitions/120.html', finding.references)
```
Comment (damiencarol) [inline dojo/unittests/tools/test_sarif_parser.py:368]: better to use `self.assertIn('http://...', finding.references)`

File: dojo/unittests/tools/test_sarif_parser.py [hunk 2]
```diff
@@ -297,25 +326,43 @@ def test_flawfinder(self):
             finding = findings[0]
             self.assertEqual("random/setstate:This function is not sufficiently random for security-related functions such as key and nonce creation (CWE-327).", finding.title)
             self.assertEqual("Medium", finding.severity)
-            self.assertEqual("This function is not sufficiently random for security-related functions such as key and nonce creation (CWE-327).", finding.description)
+            description = '''**Result message:** random/setstate:This function is not sufficiently random for security-related functions such as key and nonce creation (CWE-327).
+**Snippet:**       is.setstate(std::ios::failbit);
```
Comment (damiencarol) [inline dojo/unittests/tools/test_sarif_parser.py:330]: Snippet can break the Markdown format.
We should use triple back-tick to escape snippet of code.

File: dojo/tools/sarif/parser.py [hunk 3]
```diff
@@ -149,10 +138,92 @@ def cve_try(val):
...
+
+
+def get_description(result, rule):
+    description = ''
+    message = ''
+    if 'message' in result:
+        message = get_message_from_multiformatMessageString(result['message'], rule)
+        description += '**Result message:** {}\n'.format(message)
+    if get_snippet(result) is not None:
+        description += '**Snippet:** {}\n'.format(get_snippet(result))
```
Comment (damiencarol) [inline dojo/tools/sarif/parser.py:178]: We need to add ``` to escape the snippet code.

File: dojo/tools/sarif/parser.py [hunk 4]
```diff
@@ -149,10 +138,92 @@ def cve_try(val):
...
+            title = get_message_from_multiformatMessageString(rule['shortDescription'], rule)
+        elif title is None and 'fullDescription' in rule:
+            title = get_message_from_multiformatMessageString(rule['fullDescription'], rule)
+        elif title is None and 'name' in rule:
+            title = rule['name']
+        elif title is None:
+            title = rule['id']
+
+    if title is None:
+        title = ''
```
Comment (damiencarol) [inline dojo/tools/sarif/parser.py:154]: Not a fan of producing fake blank data.

**Review by damiencarol — COMMENTED** [review]
Many improvements. it's a good move forward.

File: dojo/unittests/tools/test_sarif_parser.py (hunk 2 above)
Comment (StefanFl) [reply dojo/unittests/tools/test_sarif_parser.py]: Very good idea, done.

File: dojo/tools/sarif/parser.py (hunk 3 above)
Comment (StefanFl) [reply dojo/tools/sarif/parser.py]: Very good idea, done.

File: dojo/tools/sarif/parser.py (hunk 4 above)
Comment (StefanFl) [reply dojo/tools/sarif/parser.py:154]: Me neither, but title cannot be None. Shall we throw an error, when there is no title?

File: dojo/unittests/tools/test_sarif_parser.py (hunk 1 above)
Comment (StefanFl) [reply dojo/unittests/tools/test_sarif_parser.py:368]: Why? This is exactly what I expect to be in the references.

Comment (damiencarol) [reply dojo/unittests/tools/test_sarif_parser.py:368]: most of the parsers use this syntax for references:
```Markdown
* [URL TITLE](https:///example.com/1)
* [URL2 TITLE](https:///example.com/2)
```
I want to keep them aligned.

File: dojo/tools/sarif/parser.py (hunk 3 above)
Comment (damiencarol) [reply dojo/tools/sarif/parser.py]: I think we need some `\n` chars to make it works as expected.

File: dojo/tools/sarif/parser.py (hunk 4 above)
Comment (damiencarol) [reply dojo/tools/sarif/parser.py:154]: I think maybe. what bother me is that the blank title could introduce some confusion for people (is it a bug (no title)?) and the code of DD (default de-duplication algorithm rely on the title I think)

File: dojo/unittests/tools/test_sarif_parser.py (hunk 1 above)
Comment (StefanFl) [reply dojo/unittests/tools/test_sarif_parser.py:368]: I still don't understand to be honest. This is a test case, it just checks that it gets exactly the value it expects from the scan file.

File: dojo/tools/sarif/parser.py (hunk 4 above)
Comment (StefanFl) [reply dojo/tools/sarif/parser.py:154]: Then I will change it to raise a `ValueError`. A lot of fields are tried to get a title. When all this fails, the report might be unusable anyway.

Comment (StefanFl) [reply dojo/tools/sarif/parser.py]: Done

Comment (damiencarol) [reply dojo/tools/sarif/parser.py]: awesome

File: dojo/unittests/tools/test_sarif_parser.py (hunk 1 above)
Comment (damiencarol) [reply dojo/unittests/tools/test_sarif_parser.py:368]: Ok just saw that we take the reference from rule data and we can't have more than one URL or a text instead. So we are not in the same case as the other parsers.

File: dojo/tools/sarif/parser.py (hunk 3 above)
Comment (StefanFl) [reply dojo/tools/sarif/parser.py]: Where do you need the `\n` characters? In one of my njsscan examples the snippet is

```
"snippet": {
    "text": "      await db.query(sql`\n        INSERT INTO users (id, email, first_name, last_name, password_hash)\n        VALUES (${join([\n          user.id,\n          user.email,\n          user.firstName,\n          user.lastName,\n          user.passwordHash,\n        ])})\n      `)"
    },
```

The description in the database is

```
**Result message:** Untrusted input concatinated with raw SQL query can result in SQL Injection.
**Snippet:** ```      await db.query(sql`
        INSERT INTO users (id, email, first_name, last_name, password_hash)
        VALUES (${join([
          user.id,
          user.email,
          user.firstName,
          user.lastName,
          user.passwordHash,
        ])})
      `)```
**Rule name:** NodeSqliInjection
```

In the UI it looks like this:
2021-09-12 16_58_39-View Finding _ DefectDojo

It looks fine to me.

Comment (damiencarol) [reply dojo/tools/sarif/parser.py]: Could you add some line feed to have the same look as for other parser (like the Dependency Check one ?)
image

Comment (StefanFl) [reply dojo/tools/sarif/parser.py]: I see, makes sense. With an additional space it looks like this now:
2021-09-12 20_35_57-View Finding _ DefectDojo

**Review by Maffooch — APPROVED** [review]
This is looking great!

File: dojo/tools/sarif/parser.py (hunk 3 above)
Comment (damiencarol) [reply dojo/tools/sarif/parser.py]: awesome!

**Review by damiencarol — APPROVED** [review]

**Pull request merged**